test/tests/functional/pbs_hook_set_jobenv.py

```diff
@@ -110,34 +109,54 @@ def setUp(self):
         self.ATTR_V = 'Full_Variable_List'
         api_to_cli.setdefault(self.ATTR_V, 'V')
 
+        # temporary files
+        (fd, fn) = tempfile.mkstemp(prefix="job_out1")
+        os.close(fd)
+        self.job_out1_tempfile = fn
+
+        (fd, fn) = tempfile.mkstemp(prefix="job_out2")
+        os.close(fd)
+        self.job_out2_tempfile = fn
+
+        (fd, fn) = tempfile.mkstemp(prefix="job_out3")
+        os.close(fd)
+        self.job_out3_tempfile = fn
+
+    def tearDown(self):
+        try:
+            os.remove(self.job_out1_tempfile)
+            os.remove(self.job_out2_tempfile)
+            os.remove(self.job_out3_tempfile)
+
+        except:
+            pass
+        finally:
+            TestFunctional.tearDown(self)
+
     def read_env(self, outputfile, ishook):
         """
         Parse the output file and store the
         variable list in a dictionary
         """
 
-        fd = open(outputfile)
-        pkey = ""
-        ppenv = {}
-        tmp = {}
-        tmp_exclude = {}
-        while(True):
-            line = fd.readline()
-            if not line:
-                break
-            l = line.split("=", 1)
-
-            if (len(l) == 2):
-                pkey = l[0]
-                if l[0] not in self.exclude_env:
-                    tmp[l[0]] = l[1]
-                    ppenv = tmp
-                else:
-                    tmp_exclude[l[0]] = l[1]
-                    ppenv = tmp_exclude
-            elif pkey != "":
-                ppenv[pkey] += l[0]
-        fd.close()
+        with open(outputfile) as fd:
+            pkey = ""
+            ppenv = {}
+            tmp = {}
+            tmp_exclude = {}
+            for line in fd:
+                l = line.split("=", 1)
+
+                if (len(l) == 2):
+                    pkey = l[0]
+                    if pkey not in self.exclude_env:
+                        tmp[pkey] = l[1]
+                        ppenv = tmp
+                    else:
+                        tmp_exclude[pkey] = l[1]
+                        ppenv = tmp_exclude
+                elif pkey != "":
+                    ppenv[pkey] += l[0]
```

Comment:
I don't understand the uses of tmp and tmp_exclude in regards to ppenv.  From the looks of it, the value of ppenv will be dependent on the last value in the file.  You go along and collect all env vars that aren't excluded in tmp.  You collect all excluded env vars in tmp_exclude.  If at any point you have a valueless env variable, you add it to the current ppenv just to overwrite ppenv back to tmp or tmp_exclude later (basically losing that value).  So at the end ppenv will either point to all the excluded vars or non-excluded vars with possibly some number of valueless env vars at the end.  It's all dependent on the last value that appears in self.exclude_env or not.

Reply:
" If at any point you have a valueless env variable, you add it to the current ppenv just to overwrite ppenv back to tmp or tmp_exclude later (basically losing that value). "
[Al] I'm appending back to 'ppenv' (line 'ppenv[pkey] += l[0]) and not overwriting it. 'ppenv' is set to the last dictionary object that was operated on, whether or not its tmp or tmp_exclude. I'm basically trying to catch environment variable lines that are long which continues to the next line. For example:

TEST_RETURN=apple,
banana,
orange

1st line: seeing <var>=<val> form, and pkey=<var>=TEST_RETURN is not in self.exclude_env list, then set tmp['TEST_RETURN'] = "apple\n", and mark ppenv = tmp so that it points to the last dictionary object that got operated on or set.
2nd line: seeing only <var> form, then goes to set ppenv[pkey] += "banana,\n", which maps to tmp['TEST_RETURN'] += "banana,\n".
3rd line: seeing only <var> form again, then goes to set ppenv[pkey] += "orange\n", which maps to tmp['TEST_RETURN'] += "orange\n".

Reply:
Ahh github strips my angled brackes. For first line, it sees VAR=VALUE form, 2nd line its VAR form only, and 3rd line its VAR form also...

Reply:
Bu the way, I found the lines "self.penv = ppenv" and "self.penv = {}" unnecessary and are actually not used. I'll remove them.

Reply:
I don't think the code does what you think it is doing.  Think of the case where you have a non-excluded var, valueless var, and then non-excluded var.  You'd add the first var to tmp and set ppenv.  You'd then add the valueless var to ppenv.  You would then add the third var to tmp and set ppenv back to tmp.  The second var is now lost.  This doesn't cover the case where you have some excluded vars and some non-excluded vars.  ppenv will only be set to either tmp or tmp_exclude at the end.  Which will be controlled by the final var you see.

Reply:
I'm confused. When you say valueless var, you mean a variable without a value, as in: "VAR="?
If so, if we have the following lines (i.e. non-excluded var, then valueless var, non-excluded var):
VAR=apple
VAR2=
VAR3=banana

results in:
tmp[VAR]=apple
tmp[VAR2]=
tmp[VAR3]=banana

When input line is split, "VAR=apple", "VAR2=", "VAR3=banana" all have line length of 2 so matching the if-condition only.
The only way to have a line length of 1 (the elif part) is if there's no '=' as in
"VAR" becomes of length 1.
I tried it interactively in Python:

>> line="VAR=VALUE"
>>> ll = line.split("=", 1)
>>> ll
['VAR', 'VALUE']
>>> len(ll)
2

>>> line2="VAR="
>>> ll2 = line2.split("=", 1)
>>> ll2
['VAR', '']
>>> len(ll2)
2

>>> line3="VAR"
>>> ll3 = line3.split("=", 1)
>>> ll3
['VAR']
>>> len(ll3)
1

Also, the for loop just collects tmp and tmp_env values, with the assumption that if there's a line that does not have '=' in it, that is, not of the form <VAR>=<VALUE>, then it will append the line to the previously entry, which could be either tmp or tmp_exclude.

In the end, the included env variables settings will be  in tmp, and excluded env settings will be in tmp_exclude.
I actually put in logger.info when I was testing and I"m seeing:

2017-09-29 20:18:52,642 INFO     ppenv = tmp
2017-09-29 20:18:52,642 INFO     line=['TEST_SQUOTE3', "the days'are here now'then\n"]
2017-09-29 20:18:52,642 INFO     <var>=<value> form
2017-09-29 20:18:52,643 INFO     line not in exclude_env so tmp[TEST_SQUOTE3]=the days'are here now'then
017-09-29 20:20:00,268 INFO     ppenv = tmp
2017-09-29 20:20:00,268 INFO     line=['TEST_RETURN', "'3,\n"]
2017-09-29 20:20:00,268 INFO     <var>=<value> form
2017-09-29 20:20:00,268 INFO     line not in exclude_env so tmp[TEST_RETURN]='3,

2017-09-29 20:20:00,268 INFO     ppenv = tmp
2017-09-29 20:20:00,268 INFO     line=['4,\n']
2017-09-29 20:20:00,268 INFO     sees on <var> form so ppenv[TEST_RETURN] += 4,

2017-09-29 20:20:00,269 INFO     line=["5'\n"]
2017-09-29 20:20:00,269 INFO     sees on <var> form so ppenv[TEST_RETURN] += 5'


2017-09-29 20:18:52,643 INFO     ppenv = tmp
2017-09-29 20:18:52,643 INFO     line=['PBS_JOBID', '182.corretja\n']
2017-09-29 20:18:52,643 INFO     <var>=<value> form
2017-09-29 20:18:52,643 INFO     line IN exclude_env so tmp_exclude[PBS_JOBID]=182.corretja

2017-09-29 20:18:52,643 INFO     ppenv = tmp_exclude
2017-09-29 20:18:52,643 INFO     line=['SSH_ASKPASS', '/usr/lib/ssh/ssh-askpass\n']
2017-09-29 20:18:52,643 INFO     <var>=<value> form
2017-09-29 20:18:52,644 INFO     line not in exclude_env so tmp[SSH_ASKPASS]=/usr/lib/ssh/ssh-askpass

And printing all the values of tmp and tmp_exclude showed:
2017-09-29 20:20:00,336 INFO     tmp[TEST_RETURN]='3,
4,
5'

2017-09-29 20:19:49,953 INFO     tmp[SSH_ASKPASS]=/usr/lib/ssh/ssh-askpass
017-09-29 20:19:49,952 INFO     tmp[TEST_SQUOTE3]=the days'are here now'then
017-09-29 20:18:52,695 INFO     tmp_exclude[PBS_JOBID]=182.corretja

Reply:
What I meant by valueless is an env var w/o an '='.  I talked with minghui and he set me straight about lists.  If you have list b and say a = b, you are setting a reference.  So you will be modifying either tmp or tmp_exclude.  This doesn't solve the problem with the final var you read will decide the final value of ppenv.  If you have 4 non-excluded vars followed by 1 excluded var, ppenv will be set to a list of 1 excluded var.  If you have the same 5 variables, but if you have the excluded var show up in the middle of the non-excluded vars, ppenv will be the 4 non-excluded vars.  What is the purpose of ppenv?  What should the value be at the end of reading all the vars in?

Reply:
ppenv is just a temporary variable for storing the reference to the last object set, whether tmp or tmp_exclude. ppenv is used when appending the continuation value to a previous value in the presence of embedded newline characters.
What 'ppenv' is set to in the end is not important. When I do "ppenv[pkey] += l[0]", I'm looking at is being equivalent to tmp[pkey] += l[0] or tmp_exclude[pkey] += l[0]. What is important is what the resulting tmp and tmp_exclude dictionaries.
" If you have 4 non-excluded vars followed by 1 excluded var, ppenv will be set to a list of 1 excluded var. If you have the same 5 variables, but if you have the excluded var show up in the middle of the non-excluded vars, ppenv will be the 4 non-excluded vars."
[Al]  What the end value of ppenv is not relevant. In here, you'll get tmp[VAR1], tmp[VAR2], tmp[VAR3], tmp[VAR4], and tmp_exclude[VAR5]. The tmp, tmp_exclude dictionaries are the ones consulted in the rest of the testsuite.

Reply:
Thanks for the explanation.  I now understand.  Maybe swap nomenclature?  What confused me is that ppenv is even more temporary than tmp_*.  Maybe rename tmp_* and name ppenv tmpenv.

Reply:
Good idea. I've renamed 'tmp' to 'penv' and 'ppenv' to 'tmpenv'.